mono/metadata/class-init.c

```diff
@@ -5127,6 +5135,179 @@ mono_class_setup_interfaces (MonoClass *klass, MonoError *error)
 	mono_loader_unlock ();
 }
 
+
+/*
+ * mono_class_setup_has_finalizer:
+ *
+ *   Initialize klass->has_finalizer if it isn't already initialized.
+ *
+ * LOCKING: Acquires the loader lock.
+ */
+void
+mono_class_setup_has_finalizer (MonoClass *klass)
+{
+	gboolean has_finalize = FALSE;
+
+	if (m_class_is_has_finalize_inited (klass))
+		return;
+
+	/* Interfaces and valuetypes are not supposed to have finalizers */
+	if (!(MONO_CLASS_IS_INTERFACE (klass) || m_class_is_valuetype (klass))) {
+		MonoMethod *cmethod = NULL;
+
+		if (m_class_get_rank (klass) == 1 && m_class_get_byval_arg (klass)->type == MONO_TYPE_SZARRAY) {
+		} else if (mono_class_is_ginst (klass)) {
+			MonoClass *gklass = mono_class_get_generic_class (klass)->container_class;
+
+			has_finalize = mono_class_has_finalizer (gklass);
+		} else if (m_class_get_parent (klass) && m_class_has_finalize (m_class_get_parent (klass))) {
+			has_finalize = TRUE;
+		} else {
+			if (m_class_get_parent (klass)) {
+				/*
+				 * Can't search in metadata for a method named Finalize, because that
+				 * ignores overrides.
+				 */
+				mono_class_setup_vtable (klass);
+				if (mono_class_has_failure (klass))
+					cmethod = NULL;
+				else
+					cmethod = m_class_get_vtable (klass) [mono_class_get_object_finalize_slot ()];
+			}
+
+			if (cmethod) {
+				g_assert (m_class_get_vtable_size (klass) > mono_class_get_object_finalize_slot ());
+
+				if (m_class_get_parent (klass)) {
+					if (cmethod->is_inflated)
+						cmethod = ((MonoMethodInflated*)cmethod)->declaring;
+					if (cmethod != mono_class_get_default_finalize_method ())
+						has_finalize = TRUE;
+				}
+			}
+		}
+	}
+
+	mono_loader_lock ();
+	if (!m_class_is_has_finalize_inited (klass)) {
+		klass->has_finalize = has_finalize ? 1 : 0;
+
+		mono_memory_barrier ();
+		klass->has_finalize_inited = TRUE;
+	}
+	mono_loader_unlock ();
+}
+
+/*
+ * mono_class_setup_supertypes:
+ * @class: a class
+ *
+ * Build the data structure needed to make fast type checks work.
+ * This currently sets two fields in @class:
+ *  - idepth: distance between @class and System.Object in the type
+ *    hierarchy + 1
+ *  - supertypes: array of classes: each element has a class in the hierarchy
+ *    starting from @class up to System.Object
+ * 
+ * LOCKING: Acquires the loader lock.
+ */
+void
+mono_class_setup_supertypes (MonoClass *klass)
+{
+	int ms, idepth;
+	MonoClass **supertypes;
+
+	mono_atomic_load_acquire (supertypes, MonoClass **, &klass->supertypes);
+	if (supertypes)
+		return;
+
+	if (klass->parent && !klass->parent->supertypes)
+		mono_class_setup_supertypes (klass->parent);
+	if (klass->parent)
+		idepth = klass->parent->idepth + 1;
+	else
+		idepth = 1;
+
+	ms = MAX (MONO_DEFAULT_SUPERTABLE_SIZE, idepth);
+	supertypes = (MonoClass **)mono_class_alloc0 (klass, sizeof (MonoClass *) * ms);
+
+	if (klass->parent) {
+		CHECKED_METADATA_WRITE_PTR ( supertypes [idepth - 1] , klass );
+
+		int supertype_idx;
```

Comment:
Does this need declared at top of scope? I can't remember if MSVC is still complaining about this?